site/docs/design-docs/021-sql-rolling-upgrade.md

```diff
@@ -0,0 +1,212 @@
+version have the same size, we choose the group with higher version; the
+reason is that members of `v+1` version will not be shut down. To
+mitigate this, the user can choose to temporarily increase the cluster
+size, the enterprise licence allows it.
+
+## Non-goals
+
+This feature won't provide rolling upgrades for Jet light jobs in
+general. To do that we would have to provide Jet client compatibility
+and DAG binary compatibility.
+
+Fault-tolerant SQL jobs also won't be supported.
+
+## Implementation details
+
+### Only implemented for light jobs
+
+We implemented this feature only for light jobs. Normal jobs can be
+fault-tolerant - to support restarting of a job on a new version, we
+would need most of the compatibility requirements listed above.
+
+### SQL operations routing
+
+SQL commands submitted from a member are optimized and submitted
+locally, but only to members with the same version as the coordinator.
+Even if it's the smaller same-version group.
+
+If the operation is sent from a client, it's routed to a member from the
+larger group. For a non-smart client, the receiving member can redirect
+the operation.
+
+### Race in member upgrades
+
+It can happen that the coordinator sends `InitExecutionOperation` to
+some member, but the member changed version in the meantime. This can
+happen because operations are routed using `Address` and the address can
+be reused by the upgraded member. To resolve this, we added
+`coordinatorVersion` parameter to `InitExecutionOperation`. The target
+will check it and throw if its version is not equal to the received
+version.
+
+### Shutdown changes
+
+During rolling upgrades, members are gracefully shut down. We want SQL
+engine to be available during this time. We'll add a new
+`JobConfig.blockShutdown` option. The SQL engine will enable this option
+for batch jobs, but not for streaming jobs.
+
+Jobs submitted during this time will not use the member in the shutdown
+process as a coordinator. Other members can handle new jobs, but they
+will not use the members being shut down. To do this, the existing
+`NotifyMemberShutdownOperation` will be broadcast to all members
+(currently it's sent just to the master).
+
+The shutdown procedure will block until all jobs with this option
+complete.
+
+Fault-tolerant jobs will be shut down immediately with a snapshot,
+regardless of this option. Other jobs (non-fault-tolerant ones without
+this option enabled) will be cancelled (light jobs) or restarted (normal
+jobs).
+
+It can happen that the member will never shut down if the user submits a
+streaming job with the `blockShutdown` option. For this scenario there
+is a hard-coded timeout of 10 seconds after which the member proceeds
+with the shutdown without waiting for all jobs to terminate. In this
+case the administrator must be able to see that job and cancel it
+manually, using either Management center or Java API. Another option is
```

Comment:
So, is it a timeout or manual cancellation?

Reply:
Both. I will finalize this as a part of the 2nd PR.